src/main/scala/io/iohk/ethereum/blockchain/sync/Blacklist.scala

```diff
@@ -0,0 +1,67 @@
+package io.iohk.ethereum.blockchain.sync
+
+import com.github.benmanes.caffeine.cache.Caffeine
+import com.github.blemale.scaffeine.{Cache, Scaffeine}
+import io.iohk.ethereum.utils.Logger
+
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.duration._
+import scala.jdk.CollectionConverters._
+import scala.jdk.OptionConverters._
+import scala.jdk.DurationConverters._
+
+import Blacklist.BlacklistId
+
+trait Blacklist {
+  def isBlacklisted(id: BlacklistId): Boolean
+  def add(id: BlacklistId, duration: FiniteDuration, reason: String): Unit
+  def remove(id: BlacklistId): Unit
+  def keys: Set[BlacklistId]
+}
+
+object Blacklist {
+  trait BlacklistId {
+    def value: String
+  }
+}
+
+final case class CacheBasedBlacklist(cache: Cache[BlacklistId, String]) extends Blacklist with Logger {
```

Comment:
Is string value (being a reason for blacklisting) ever used? Looks like we are not interested in it. Maybe for space efficiency we could use Boolean or a constant value? Or maybe there's a set implementation of cache available?

Reply:
At the moment it isn't used but I think it would be a very interesting metric to add to our dashboard. I turned the reason Strings into proper types and the different types of reasons are now represented as objects so it shouldn't have a negative impact on memory.

Reply:
good point on adding a metric and seeing which reasons for blacklisting prevail. So thumbs up for introducing types for blacklisting reasons, but I think they are descriptive enough, so we don't have to keep the record for extra description. I'll leave another comment with proposed simplification

Reply:
Replied on the other comment but I disagree. The name does not contain the additional information that is currently available for debugging.